### PR TITLE
Split test and bench out to separate packages

### PR DESCRIPTION
**Pull request by wraithm**
https://github.com/haskell-bitcoin/bitcoin/issues/42

### CODE REVIEW

File: package.yaml
Comment (wraithm) [inline package.yaml:50]: This is maybe a questionable thing, but tbh, I really dislike `other-modules` as a thing. Like, what if I want the fingerprint's `Word32` or whatever. Just lemme do it.

File: benchmark/package.yaml
Comment (wraithm) [inline benchmark/package.yaml]: I need to prune these deps.

Comment (wraithm) [reply benchmark/package.yaml]: fixed

**Comment by GambolingPangolin** [comment]
The packaging is not very clear from the directory structure.  How about reorganizing the code so that there are three folders, each of which contains a package?:

- `bitcoin`
- `bitcoin-test`
- `bitcoin-bench`

In addition, we can keep `README.md`, `stack.yaml`, etc. at the top level.

File: package.yaml
Comment (GambolingPangolin) [reply package.yaml:50]: I do not like representing `Fingerprint` as a `Word32`.  How about we just change it to `SizedByteArray 4 ByteString`?  Then I would be more comfortable exposing `Fingerprint` internals, and maybe getting rid of `Bitcoin.Keys.Extended.Internal`.

Comment (wraithm) [reply package.yaml:50]: IMHO, that sounds out of scope for this changeset, but certainly an interesting thought.

**Comment by wraithm** [comment]
> The packaging is not very clear from the directory structure. How about reorganizing the code so that there are three folders, each of which contains a package?:
> 
> * `bitcoin`
> * `bitcoin-test`
> * `bitcoin-bench`
> 
> In addition, we can keep `README.md`, `stack.yaml`, etc. at the top level.

I'm not opposed to that at all.

@ProofOfKeags, what do you think?

File: package.yaml
Comment (ProofOfKeags) [reply package.yaml:50]: Agreed different changeset. I think exposing Internal modules is common practice in our ecosystem. People know the footguns they entail, let's just expose it.

**Comment by ProofOfKeags** [comment]
100% onboard with that folder structure.

**Comment by ProofOfKeags** [comment]
After reviewing as well I'm fine to merge this after exposing internal modules and reorganizing the directories to be more intuitive.

**Comment by wraithm** [comment]
@ProofOfKeags @GambolingPangolin This is ready for review.

File: bitcoin-test/package.yaml
Comment (GambolingPangolin) [inline bitcoin-test/package.yaml:19]: This should prob have a version, same for the bench package if applicable.

Comment (ProofOfKeags) [reply bitcoin-test/package.yaml:19]: Why, shouldn't it always be in sync with the one in the repo?

Comment (GambolingPangolin) [reply bitcoin-test/package.yaml:19]: It depends on if we want to support users directly depending on the test package.  I have found that it generally helps to have a lot of helper functions and maybe even a framework for testing bitcoin.  We could treat `bitcoin-test` as a product, which exports that kind of stuff.  It would be too opinionated to export directly from `bitcoin`, but that's ok.  However, if we just want this package to be the unit tests, then I agree a version is not necessary.

After writing this note, I do think that `bitcoin-test` should just be for unit tests, so no version necessary.

**Review by GambolingPangolin — APPROVED** [review]

File: bitcoin-test/package.yaml
Comment (ProofOfKeags) [reply bitcoin-test/package.yaml:19]: so let's send it?

**Pull request merged**
